numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1437 @@
+ * discovery can be implemented sufficiently fast.  That is, it is not
+ * necessary to create fast paths that only find the correct shape e.g. when
+ * ``dtype=np.dtype("f8")`` is given.
+ *
+ * The code here avoid multiple conversion of array-like objects (including
+ * sequences). These objects are cached after conversion, which will require
+ * additional memory, but can drastically speed up coercion from from array
+ * like objects.
+ */
+
+
+/*
+ * For finding a DType quickly from a type, it is easiest to have a
+ * a mapping of pytype -> dtype.
+ * Since a DType must know its type, but the type not the DType, we will
+ * store the DType as a weak reference. When a reference is dead we can
+ * remove the item from the dictionary.
+ * A cleanup should probably be done occasionally if (and only if) a large
+ * number of type -> DType mappings are added.
+ * This assumes that the mapping is a bifurcation DType <-> type
+ * (there is exactly one DType for each type and vise versa).
+ * If it is not, it is possible for a python type to stay alive unnecessarily.
+ */
+PyObject *_global_pytype_to_type_dict = NULL;
+
+
+/* Enum to track or signal some things during dtype and shape discovery */
+enum _dtype_discovery_flags {
+    FOUND_RAGGED_ARRAY = 1,
+    GAVE_SUBCLASS_WARNING = 2,
+    PROMOTION_FAILED = 4,
+    DISCOVER_STRINGS_AS_SEQUENCES = 8,
+    DISCOVER_TUPLES_AS_ELEMENTS = 16,
+    MAX_DIMS_WAS_REACHED = 32,
+    DESCRIPTOR_WAS_SET = 64,
+};
+
+
+/**
+ * Adds known sequence types to the global type dictionary, note that when
+ * a DType is passed in, this lookup may be ignored.
+ *
+ * @return -1 on error 0 on success
+ */
+static int
+_prime_global_pytype_to_type_dict()
+{
+    int res;
+
+    /* Add the basic Python sequence types */
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyList_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyTuple_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    /* NumPy Arrays are not handled as scalars */
+    res = PyDict_SetItem(_global_pytype_to_type_dict,
+                         (PyObject *)&PyArray_Type, Py_None);
+    if (res < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+
+/**
+ * Add a new mapping from a python type to the DType class.
+ *
+ * This assumes that the DType class is guaranteed to hold on the
+ * python type (this assumption is guaranteed).
+ * This functionality supercedes ``_typenum_fromtypeobj``.
+ *
+ * @param DType DType to map the python type to
+ * @param pytype Python type to map from
+ * @param userdef Whether or not it is user defined. We ensure that user
+ *        defined scalars subclass from our scalars (for now).
+ */
+NPY_NO_EXPORT int
+_PyArray_MapPyTypeToDType(
+        PyArray_DTypeMeta *DType, PyTypeObject *pytype, npy_bool userdef)
+{
+    PyObject *Dtype_obj = (PyObject *)DType;
```

Comment:
initially i was confused if this was a dtype class or a dtype object. would Dtype_class or Dtype_type make more sense here (probably with a _obj suffix) ?

Reply:
Hmmm, I think in most cases I use `descr` currently for the instance and often probably upper case for the class, maybe just making sure it is almost always upper-case for the class? And `descr` in any place its the instance but its unclear?

Reply:
since ``descr`` is the convention for instance this is probably fine the way it is.